File: 449. Serialize and Deserialize BST/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return ""
        
        postorder = []
        stack = []
        node = root
        left = 1
        right = 2
        while node or stack:
            if node:
                stack.append((node, left))
                node = node.left
            else:
                node, status = stack.pop()
                if status == left:
                    stack.append((node, right))
                    node = node.right
                else:
                    postorder.append(str(node.val))
                    node = None
        return "#".join(postorder)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if not data:
            return None
        
        postorder = data.split("#")
        root = TreeNode(int(postorder.pop()))
        stack = [root]
        while postorder:
            node = TreeNode(int(postorder.pop()))
            if node.val > stack[-1].val:
                stack[-1].right = node
            else:
                parent = stack.pop()
                while stack and stack[-1].val > node.val:
                    parent = stack.pop()
                    
                parent.left = node
                
            stack.append(node)
            
        return root
# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: 449. Serialize and Deserialize BST/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) {
            return new String("");
        }
        
        Stack<TreeNode> stack = new Stack<TreeNode>();
        List<String> sol = new ArrayList<String>();
        while (root != null || !stack.isEmpty()) {
            if (root != null) {
                sol.add(String.valueOf(root.val));
                stack.push(root.right);
                root = root.left;
            } else {
                while (!stack.isEmpty() && root == null) {
                    root = stack.pop();
                }
            }
        }
        
        return String.join(",", sol);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.equals("")) {
            return null;
        }
        
        List<String> sol = new ArrayList<String>(Arrays.asList(data.split(",")));
        Collections.reverse(sol);
        TreeNode root = new TreeNode(Integer.parseInt(sol.remove(sol.size()-1)));
        TreeNode node, parent;
        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.push(root);
        while (sol.size() > 0) {
            node = new TreeNode(Integer.parseInt(sol.remove(sol.size()-1)));
            parent = stack.pop();
            if (parent.val > node.val) {
                parent.left = node;
                stack.push(parent);
            } else {
                while (!stack.isEmpty() && stack.peek().val < node.val) {
                    parent = stack.pop();
                }
                parent.right = node;
            }
            stack.push(node);
        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));
